### DISCUSSION TITLE
Remap/swap characters on the same key

### DISCUSSION THREAD

**Discussion by gerardbosch** [kmonad/kmonad#754] [Q&A]
Hi! I asked this a looong time ago, but I'm not sure if KMonad can remap characters:

* Is it possible to swap 2 key symbols for a key? I want to swap characters `;` and `:` on US layout. So hitting the button straight produces `:` and using shift produces `;` (the reverse of the default).

In case KMonad cannot remap chars in this way, which tool can I use, prefearably multi platform? I used XKB on Linux for this, but I faced problems, and looking for something better. I'm already using KMonad BTW, so I'd like to see if I can accomplish this goal with it 🙂 —or if not, a simple/effective/multiplatform alternative to do it.

Thanks!

**Comment by ArielHorwitz** [comment]
Didn't test this, but I think something like this will work:
```
(defcfg
  input...
  output...
  fallthrough true
)

(defsrc
  esc   f1    f2    f3    f4    f5    f6    f7    f8    f9    f10   f11   f12
  grv   1     2     3     4     5     6     7     8     9     0     -     =     bspc
  tab   q     w     e     r     t     y     u     i     o     p     [     ]     \
  caps  a     s     d     f     g     h     j     k     l     ;     '     ret
  lsft  z     x     c     v     b     n     m     ,     .     /     rsft
  lctl  lmet  lalt              spc               ralt  rmet  cmp   rctl
)

(deflayer base
  _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     :     _     _
  @flip _     _     _     _     _     _     _     _     _     _     _
  _     _     _                 _                 _     _     _     _
)
(def alias
    flip (around lsft (layer-toggle flipped_colon))
)

(deflayer flipped_colon
  _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     ;     _     _
  _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _                 _                 _     _     _     _
)
```

Let me know if this helped.

**Reply by gerardbosch** [reply]
Hi @ArielHorwitz, thanks for your answer!

I tired your suggestion but it does not work. When I do this, the base layer do produces the character I want (`:`), but your `flipped_colon` layer produces also the same char (`:`), disregarding I place in the "shifted layer" the semicolon `;` or the colon `:`.

I tried to put another character in the "shifted layer" to see what happens, and if I put an `a` in the colon position, what I get is `A`. If I put an `A`, I also get an `A`.

So I don't know how to force it to emit the `;` 😟

**Reply by ArielHorwitz** [reply]
Ah I think I see the issue. It may be because of the `around lsft`. The reason I put it there was to allow all the `_` characters to "fallthrough" and still get all the capital letters. But it may work if you remove the `around lsft` and manually add each capital character in the `flipped_colon` layer.

Sorry I don't have the time to test -- I can try over the weekend.

**Reply by gerardbosch** [reply]
The _fallthrough_ idea sounded good, as redefining the whole shifted layer for every single key looks an overkill, very verbose just to swap a single key or few keys. But maybe KMonad does not offer a straightforward way to achieve it, I don't know, maybe I should explore another tool for this, but I don't find the right one.

Anyway, thank you and please, let me know if you manage to make it work 🙂

**Reply by ArielHorwitz** [reply]
Ok just tested because I was too curious (though my setup is a bit complicated so I can't be sure if I tested it properly) and this should work, although it is a little verbose:

```
(deflayer base
  _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     :     _     _
  @flip _     _     _     _     _     _     _     _     _     _     @flip
  _     _     _                 _                 _     _     _     _
)
(deflayer flipped_colon
  _     _     _     _     _     _     _     _     _     _     _     _     _
  _     _     _     _     _     _     _     _     _     _     _     _     _     _
  _     Q     W     E     R     T     Y     U     I     O     P     _     _     _
  _     A     S     D     F     G     H     J     K     L     ;     _     _
  _     Z     X     C     V     B     N     M     _     _     _     _
  _     _     _                 _                 _     _     _     _
)
(defalias
    flip (layer-toggle flipped_colon)
)
```

Let me know if it did for you.

**Reply by gerardbosch** [reply]
I made a quick test and this would break the normal shift behavior. I would need to redefine the whole layer, including numbers and symbols, otherwise, pressing `sft-2` produces `2` intead of `@`.

But the problem is not this, the thing is that the Shift key becomes a "ghost", as now, shift key only activates a layer. If I run `xev` to debug it, when I press and release the shift, it prints nothing, when I would expect `keycode 50 (keysym 0xffe1, Shift_L)`. So this makes shift effectively inexistent, and any system (or app) shortcut based on shift would not work, e.g. `Shift-F4`.

Thanks a lot for your support!! :) I'll leave the question open until we can have an authoritative answer about if this kind of key remapping is possible in KMonad, or to hear the most appropriate tool.

**Reply by gerardbosch** [reply]
Hi! I finally switched from XKB to `keyd` —I managed to make it work together with KMonad. Still not a cross-platform solution, but it was simple and does the job. See here: https://github.com/rvaiya/keyd/issues/289#issuecomment-1846190796

I leave the question open, to see if there was a simple way to achieve it with KMonad or not. Thanks!
